OAK-9566 | Improve index stats

nit0906 opened the pull request:
* Added support for Json format 
* Fixed issue with local index dir size - currently with composite node store - it was just reporting size of read only mount
* Added new stats for indexCreationTimestamp, ReindexCompletionTimestamp, suggestDirSize, If the index has hidden property index node and if it has a hidden read only libs mount.

Commit:
OAK-9566 | Improve index stats

Commit:
Removing unused and star imports

Commit:
minor formatting fix

fabriziofortino commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -46,7 +47,8 @@
             "felix.inventory.printer.title=Oak Index Stats",
-            "felix.inventory.printer.format=TEXT"
+            "felix.inventory.printer.format=TEXT",
+                "felix.inventory.printer.format=JSON"
```
minor: align formatting

nit0906 replied:
done

fabriziofortino commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,193 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
+    }
+
+    // Wrappers around JsonBuilder that will result in NOOP if builder is null -
+    // These are just to avoid the mulitple if/else check while handling for both TEXT and JSON formats
```
typo
```suggestion
    // These are just to avoid the multiple if/else check while handling for both TEXT and JSON formats
```

nit0906 replied:
done

fabriziofortino commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,193 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
             AsyncIndexInfo info = asyncIndexInfoService.getInfo(lane);
             if (info != null) {
-                        pw.printf("    Last indexed to      : %s%n", formatTime(info.getLastIndexedTo()));
+                startJsonObject(json);
```
`startJsonObject(json);` is called in both `if` and `else` blocks. You can remove from both an move it before the `if` block.

nit0906 replied:
done

fabriziofortino commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,193 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
-            if (diff != null) {
-                pw.println("    "+diff);
-            }
+        if (info.getType() == "lucene") {
```
use `equals` instead of `==`

nit0906 replied:
done

fabriziofortino commented on oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinterTest.java:
```diff
@@ -49,11 +53,30 @@ public void asyncIndexInfo() throws Exception {
+        Map<String, JsonObject> jsonMap = json.getChildren();
+        assertTrue(jsonMap.keySet().contains("Async Indexers State"));
+        int size  = Integer.parseInt(jsonMap.get("Async Indexers State").getProperties().get("Number of async indexer lanes"));
+        assertTrue(size == 2);
```
```suggestion
        assertEquals(2, size);
```

nit0906 replied:
done

Commit:
Review comments

thomasmueller reviewed (commented):
Could you add an example output, in Json format, to https://issues.apache.org/jira/browse/OAK-9566 ?
Or (might be easier) add a test case with Json output... anyway having a test case would be good.

fabriziofortino commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,190 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
-            if (diff != null) {
-                pw.println("    "+diff);
-            }
+        if (info.getType().equals("lucene")) {
```
just to avoid npe in case the type is not specified
```suggestion
        if ("lucene".equals(info.getType())) {
```

nit0906 replied:
done

thomasmueller commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,190 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
+                json.value((Boolean)value);
+            } else if (value instanceof Integer) {
+                json.value((Integer) value);
+            }
```
I would probably handle "null" separately at the beginning (if value == null). Could throw an exception.
Also, "else" for all other cases is missing (could e.g. throw an exception)

nit0906 replied:
done

thomasmueller commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,190 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
-                printIndexInfo(pw, info);
+                printIndexInfo(pw, json, info, format);
+                if (format == Format.TEXT) {
+                    if (info.hasIndexDefinitionChangedWithoutReindexing()) {
```
The json output could also have this info I assume, right?

nit0906 replied:
yes - I was thinking that this diff was also in json, and might cause formatting issues - but it's not in json. I have modified for the json stats also to include this

nit0906 commented:
@thomasmueller -
>> Could you add an example output, in Json format, to https://issues.apache.org/jira/browse/OAK-9566 ?

Attached a sample o/p to the jira. Please note that the values in it are not correct and some are missing, since It's from an earlier revision and was not run on composite node store - but the format is same so it'll give an idea about how it's structured.

 >> Or (might be easier) add a test case with Json output... anyway having a test case would be good.

I added a couple of tests cases. See IndexPrinterTest#asyncIndexInfoJson and  IndexPrinterTest#indexInfoJson

Commit:
minor

thomasmueller commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +68,190 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
 
         if (info.getSizeInBytes() >= 0){
-            pw.printf("    Size                    : %s%n", IOUtils.humanReadableByteCount(info.getSizeInBytes()));
+            keyValue("    Size                     ", IOUtils.humanReadableByteCount(info.getSizeInBytes()), pw, json, format);
```
My personal opinion is that we should list all size data in the same way, e.g. all in GB. Examples:
* 1000 (= 1 TB)
* 1 (= 1 GB)
* 0.001 (= 1 MB)
* 0.000001 (= 1 B)

nit0906 replied:
@thomasmueller  - I disagree a bit here. Since this is human readable format (that would not be consumed by any code for calculations/alerting ), so I think, 1 B makes more sense than 0.000001 GB. 

If we want to use these for some calculations/alerting - then I agree with your point that they should be in the same unit (and in which case we would have to change the formatting even a bit more ), because here the o/p would be string - 10B, 10MB, 10GB and like that.

thomasmueller replied:
> I disagree a bit here.

That's fine, it's just my personal opinion. When I investigate a problem, in the Lucene index stats, it I usually search for "GB" and then look at only those that have this unit; or use grep "GB". I would miss entries with "TB" however. 

So I argue that even for a human, using the same unit would make sense, assuming there is a sensible default unit.

We could round to 3 digits after the comma (e.g. round up from 0.000001 GB => 0.001 GB).

thomasmueller replied:
> If we want to use these for some calculations/alerting

Looking at the json output, I think it would make sense to use it for alerting... What about we list both the "human readable" as well as the "same unit" numbers?

nit0906 replied:
Makes sense. I will add both the String format (that appends  B, MB, GB  accordingly) and the numeric one (in bytes).

Commit:
review comments

thomasmueller commented:
I wonder if the timestamps would be better listed in UTC? Right now we have

    "Last indexed up to": "2021-09-17T12:52:02.619+05:30",

That would become

    "Last indexed up to": "2021-09-17T07:52:02.619Z",

Where the "Z" is from here: https://www.w3.org/TR/NOTE-datetime

nit0906 commented:
@thomasmueller 
>>> I wonder if the timestamps would be better listed in UTC?

Sure - I transformed them in UTC now. And also added another property to show the sized in bytes (numeric) that can be used for alerts etc. Added a new sample json here - https://issues.apache.org/jira/browse/OAK-9566?focusedCommentId=17417733&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17417733 .

Commit:
Making timestamp in UTC and adding another property to show size in bytes

Commit:
Minor , remove comment

Commit:
Merge branch 'trunk' into OAK-9566_1

thomasmueller commented:
> https://issues.apache.org/jira/browse/OAK-9566?focusedCommentId=17417733&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17417733

The example uses "2021-55-20T16:55:53.53 UTC", you have used seconds instead of month...

While "2021-09-20T16:55:53.53 UTC" is clear for the user, is not a standardized format... https://www.w3.org/TR/NOTE-datetime specifies "2021-09-17T07:52:02.619Z". 

If we want "human readable", then what about "2021-09-20 16:55:53 UTC" (no "T", and no fractional seconds).

thomasmueller commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +71,200 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
         cal.setTimeInMillis(time);
-        return ISO8601.format(cal);
+        Date date = cal.getTime();
+        SimpleDateFormat outputFmt = new SimpleDateFormat("yyyy-mm-dd'T'HH:mm:ss.s z");
```
mm is for minutes, but you need month, I think that would be MM. "ss.s" is incorrect, it shows seconds twice. I don't think we need milliseconds or fractional seconds, but if you want them then this needs to be fixed.

For human readable, what about

    SimpleDateFormat outputFmt = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss 'UTC'");

For ISO 8601, I think this would be correct:

    SimpleDateFormat outputFmt = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");

nit0906 replied:
Changed the format to look like this now 
`2021-09-21T07:16:53 UTC`

nit0906 replied:
`2021-09-21T07:16:53Z` Set format to 
`SimpleDateFormat outputFmt = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");`

thomasmueller commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/inventory/IndexPrinter.java:
```diff
@@ -66,97 +71,200 @@ public IndexPrinter(IndexInfoService indexInfoService, AsyncIndexInfoService asy
+    private static void keyValue(String key, Object value, PrintWriter pw, JsopBuilder json, Format format) {
+        // In case the key is null or an emppty String or value is null,
+        // throw an IllegalArgumentException.
+        if (value == null || key == null || key.equals("")) {
```
I would support null values.

nit0906 replied:
@thomasmueller 
I actually added the check post your comment earlier on - 

> I would probably handle "null" separately at the beginning (if value == null). Could throw an exception.

I can remove it .

nit0906 replied:
done

Commit:
fixing the date format
